packages/opentelemetry-plugin-mysql/src/mysql.ts

```diff
@@ -0,0 +1,266 @@
+/*!
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { BasePlugin } from '@opentelemetry/core';
+import { CanonicalCode, Span, SpanKind } from '@opentelemetry/types';
+import * as mysqlTypes from 'mysql';
+import * as shimmer from 'shimmer';
+import { AttributeNames } from './enums';
+import { getConnectionAttributes, getSpanName } from './utils';
+
+export class MysqlPlugin extends BasePlugin<typeof mysqlTypes> {
+  readonly moduleName = 'mysql';
+  readonly supportedVersions = ['2.*'];
+
+  static readonly COMPONENT = 'mysql';
+  static readonly DB_TYPE = 'sql';
+
+  static readonly COMMON_ATTRIBUTES = {
+    [AttributeNames.COMPONENT]: MysqlPlugin.COMPONENT,
+    [AttributeNames.DB_TYPE]: MysqlPlugin.DB_TYPE,
+    [AttributeNames.PEER_SERVICE]: MysqlPlugin.COMPONENT,
+  };
+
+  protected patch(): typeof mysqlTypes {
+    shimmer.wrap(
+      this._moduleExports,
+      'createConnection',
+      this._patchCreateConnection() as any
+    );
+
+    shimmer.wrap(
+      this._moduleExports,
+      'createPool',
+      this._patchCreatePool() as any
+    );
+
+    shimmer.wrap(
+      this._moduleExports,
+      'createPoolCluster',
+      this._patchCreatePoolCluster() as any
+    );
+
+    return this._moduleExports;
+  }
+
+  protected unpatch(): void {
+    shimmer.unwrap(this._moduleExports, 'createConnection');
+    shimmer.unwrap(this._moduleExports, 'createPool');
+    shimmer.unwrap(this._moduleExports, 'createPoolCluster');
+  }
+
+  // global export function
+  private _patchCreateConnection() {
+    return (originalCreateConnection: typeof mysqlTypes.createConnection) => {
+      const thisPlugin = this;
+      thisPlugin._logger.debug(
+        'MysqlPlugin#patch: patched mysql createConnection'
+      );
+
+      return function createConnection(
+        _connectionUri: string | mysqlTypes.ConnectionConfig
+      ) {
+        //@ts-ignore
+        const originalResult = originalCreateConnection(...arguments);
+
+        shimmer.wrap(
```

Comment:
When do we unwrap this function?

Reply:
I don't want to keep a hard reference to objects generated by calling `createConnection` because if they go out of scope in the user's code, I don't want us to be the memory leak. Imagine code like this:

```typescript
app.get("/user/:id", (req, res) => {
  const conn = mysql.createConnection(/* connection params */);

  conn.query("SELECT * FROM Users WHERE id = ?", [req.params.id], (err, result) => {
    const user = getUser(result);
    res.json(user);
  });
});

```

If we hold a hard reference to that connection in order to unwrap it, it will never be garbage collected. A new connection will be held in memory indefinitely every time the route is called.

Reply:
@mayurkale22 take a look at this solution and LMK what you think:

https://github.com/dynatrace-oss-contrib/opentelemetry-js/commit/5089692b7409bac562c235e8b07a37ea2fcb3efc